### DISCUSSION TITLE
Custom JWK provisioner

### DISCUSSION THREAD

**Discussion by himdeeppathak** [smallstep/certificates#1234] [Q&A]
I want to add a new custom JWK provider having only public key part in CA side. private key part I want to keep in remote step client side. For that I followed these steps:

I created JWK key sets at client side using 
`step crypto jwk create pub.json priv.json`
I added new JWK provisioner with created public key
`step ca provisioner add jane@doe.com --type JWK --public-key pub.json` 

Now I want to create a new certificate at remote step client side using custom JWK provisioner using 
`step ca certificate <subject> <crt-file> <key-file>`

I can create signed JWT using 
JWT=$(step crypto jwt sign --key priv.json --aud foo --iss bar --sub foobar --exp $(date -d'+5minutes' +%s))
Now I am stuck how to pass on custom JWK provider and  signed JWT to step ca certificate command
Is this supported?

**Comment by tashian** [comment]
@himdeeppathak, in this case you can pass `--token $JWT --provisioner jane@doe.com` to `step ca certificate`.

**Reply by himdeeppathak** [reply]
Unfortunately `step ca certificate <subj> <cert> <key>--token $JWT  --provisioner <name>` does not support JWT token. 
I could solve it using `step ca sign <csr> <cert>  --token $JWT..`

**Reply by maraino** [reply]
This works, the provisioner flag is not necessary if you pass the token, but it works too. An easy way to demonstrate is is using a provisioner with the private part, I can use `step ca token` to create the token, and then pass the flag:

```console
$ export JWT=`step ca token mariano@smallstep.com`
✔ Provisioner: mariano@smallstep.com (JWK) [kid: nvgnR8wSzpUlrt_tC3mvrhwhBx9Y7T1WL_JjcFVWYBQ]
Please enter the password to decrypt the provisioner key:
$ step ca certificate mariano@smallstep.com mariano.crt mariano.key --token $JWT
✔ CA: https://ca.smallstep.com:9000
✔ Certificate: mariano.crt
✔ Private Key: mariano.key
```

If I build the `$JWT` manually and then get rid of the `encryptedKey` works as expected too. What it won't work is re-using the `$JWT`.

**Reply by himdeeppathak** [reply]
I completely agree.  
My use case is bit different - not to share private key part with CA server side when using a custom JWK provisioner, create the signed token at client side and send to sever for validation.

**Reply by maraino** [reply]
I'm not sure if I understand this.

> not to share private key part with CA server side when using a custom JWK provisioner.

You can do that by removing the `encryptedKey` in the `ca.json` or just passing `--public-key` and omitting `--private-key` if you create a provisioner using the admin API.

> create the signed token at client side

Ok, you can create it with whatever means you want `./mytoken.sh`, request to a remote server, ...

> and send to sever for validation

To validate a JWT the server only needs the public key.
